fuzz/writeout.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (C) 2024 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "writeout.h"
+
+
+static const char* ycbcr_to_str(ASS_YCbCrMatrix ycbcr_val)
+{
+    switch (ycbcr_val) {
+        case YCBCR_NONE:         return "None";
+        case YCBCR_BT601_TV:     return "TV.601";
+        case YCBCR_BT601_PC:     return "PC.601";
+        case YCBCR_BT709_TV:     return "TV.709";
+        case YCBCR_BT709_PC:     return "PC.709";
+        case YCBCR_SMPTE240M_TV: return "TV.240m";
+        case YCBCR_SMPTE240M_PC: return "PC.240m";
+        case YCBCR_FCC_TV:       return "TV.fcc";
+        case YCBCR_FCC_PC:       return "PC.fcc";
+        case YCBCR_DEFAULT:      return "ThereWasNoHeader";
+        case YCBCR_UNKNOWN:
+        default:
+            return "InvalidUnknownValue";
+    }
+}
+
+static bool time_to_str(long time, char** buf, size_t* buf_size)
+{
+    time /= 10; // ASS files can only have centi-second precision
+    int sign = time < 0 ? -1 : 1;
+    time = labs(time);
+
+    // Without chars for hours, except for sign indicator
+    size_t min_size = sign == -1 ? 14 : 10;
+    int cs, s, m;
+    #define DIVMOD(div,mod) mod = time % div; time /= div
+    DIVMOD(100, cs);
+    DIVMOD(60, s);
+    DIVMOD(60, m);
+    #undef DIVMOD
+
+    min_size += time >= 10 ? ceil(log10(time + 1)) : 1;
```

Comment:
it turns out with a common 64-bit `long` this gives `+= 13` so in practice the later “`+14` arbitrary padding” already prevented more than two reallocs from ever needing to happen on such platforms.With the actual initial size of 128 used in `write_events` no reallocs occur in practice. There’s a (probably theoretical) catch due to differences in absolute value between `LONG_MAX` and `LONG_MIN`, but this is solvable

Can you explain more what you have in mind for assuming a fixed-size buffer?

Afaict for being able to pass in a guaranteed safe, fixed-size buffer and never needing to do any size checks or allocs here, I think we’d need to (a) just trust callers or (b) hardcode the ceiled log10 value *(with an assert on `LONG_MAX`; even C23 constexpr doesn't help due to needing a function call for the logarithm)* neither of which i particularly like

Still checking a passed-in size parameter and just failing instead of reallocing if the buffer is too small would work ig, but is this really desirable or a notable simplification?

Reply:
Actually, `long` type is incorrect here, should be `long long` or `int64_t`. And maximal value of `time` is `3661010 * (long long) INT32_MAX`, see [here](https://github.com/libass/libass/blob/master/libass/ass.c#L229-L240). So, we can calculate maximum length based on that value. Or assume maximum 64-bit or even 128-bit value, not really a big difference.

Fixed buffer without caller trust can be implemented with
```c
typedef struct {
     char value[64];
} TimeBuffer;
```

Reply:
> Actually, `long` type is incorrect here, should be `long long`

yep, good catch

> Fixed buffer without caller trust can be implemented with

This still hardcodes the assumed to be safe value derived from `log10`


> And maximal value of time is 3661010 * (long long) INT32_MAX, see [here](https://github.com/libass/libass/blob/master/libass/ass.c#L229-L240).

Right, we changed this to explicit 32-bit components recent'ish. In that case, there’s no risk of the max value growing further on other platforms and  hardcoding miiight be fine.

Reply:
Pushed a fix for `long long` and a separate commit for fixed-size buffers

Reply:
> (b) hardcode the ceiled log10 value _(with an assert on `LONG_MAX`; even C23 constexpr doesn't help due to needing a function call for the logarithm)_

As little as C23 constexpr may currently matter, would this not work?

```c
constexpr size_t maxlen = (int) (CHOSENTYPENAME_WIDTH * 0.30102999566398119521373889472449 + 1);
```

Reply:
yeah you’re right. With constexpr we could just hardcode the platform-independent `log10(2)` and use `log2(LLONG_MAX)  ≤ sizeof(long long) * CHAR_BITS)`